File: C/0016.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

// Comparison function for qsort
int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int threeSumClosest(int* nums, int numsSize, int target) {
    if (numsSize < 3) return 0;

    // Sort the array
    qsort(nums, numsSize, sizeof(int), compare);
    
    long closestSum = INT_MAX;

    for (int i = 0; i < numsSize - 2; i++) {
        int left = i + 1;
        int right = numsSize - 1;

        while (left < right) {
            int sum = nums[i] + nums[left] + nums[right];
            
            // Update the closest sum if the current sum is closer to the target
            if (abs(target - sum) < abs(target - closestSum)) {
                closestSum = sum;
            }
            
            // Move the pointers
            if (sum < target) {
                left++;
            } else if (sum > target) {
                right--;
            } else {
                // If the sum is exactly the target, we have found the closest possible sum
                return sum;
            }
        }
    }
    
    return closestSum;
}

File: Scala/0016.scala
object Solution {
    def threeSumClosest(nums: Array[Int], target: Int): Int = {
        // Sort the array
        val sortedNums = nums.sorted
        
        // Initialize the closest sum with a large value
        var closestSum = Int.MaxValue
        
        // Iterate through the array
        for (i <- sortedNums.indices) {
            // Use two pointers for the remaining elements
            var left = i + 1
            var right = sortedNums.length - 1
            
            while (left < right) {
                val sum = sortedNums(i) + sortedNums(left) + sortedNums(right)
                
                // Update closestSum if the current sum is closer to the target
                if (Math.abs(sum - target) < Math.abs(closestSum - target)) {
                    closestSum = sum
                }
                
                // Move pointers based on comparison with target
                if (sum < target) {
                    left += 1
                } else if (sum > target) {
                    right -= 1
                } else {
                    // If sum exactly matches the target, update closestSum and exit the loop
                    closestSum = sum
                    // Break out of the inner while loop
                    left = right // This will exit the loop as left >= right
                }
            }
        }
        
        // Return the closest sum found
        closestSum
    }
}
